os/kernel/binary_manager/binary_manager.c

```diff
@@ -0,0 +1,375 @@
+		return ERROR;
+	}
+
+	/* Read the number of binaries */
+	ret = read(fd, (FAR uint8_t *)&bin_count, sizeof(uint16_t));
+	if (bin_count <= 0) {
+		bmvdbg("Invalid config file!\n");
+		close(fd);
+		return ERROR;
+	}
+
+	/* Read the binary config file */
+	for (idx = 0; idx < bin_count; idx++) {
+		memset(&config_data, 0, sizeof(binary_config_t));
+		ret = read(fd, (FAR uint8_t *)&config_data, sizeof(binary_config_t));
+		if (ret == sizeof(binary_config_t)) {
+			bmvdbg("Binary config : %s %d\n", config_data.name, config_data.use_flag);
+			/* The flag should be 0 or 1 value because each binary uses two partitions */
+			if (config_data.use_flag < 0 || config_data.use_flag >= 2) {
+				bmvdbg("Invalid config value! %d\n", ret);
+				continue;
+			}
+			/* Search binary and set flag */
+			for (idx = 0; idx < g_bin_count; idx++) {
+				if (!strncmp(BIN_NAME(idx), config_data.name, strlen(config_data.name) + 1)) {
+					BIN_USEIDX(idx) = config_data.use_flag;
+					bmvdbg("FIND!!! %s flag %d \n", BIN_NAME(idx), BIN_USEIDX(idx));
+					break;
+				}
+			}
+		} else {
+			bmvdbg("Invalid config file! %d\n", ret);
+			break;
+		}
+	}
+	close(fd);
+
+	return OK;
+}
+
+static void binary_manager_load_bininfo(void)
+{
+	int ret;
+	int bin_idx;
+	binary_header_t header_data;
+
+	ret = binary_manager_read_config();
+	if (ret < 0) {
+		return;
+	}
+
+	/* Read header of all binaries */
+	for (bin_idx = 0; bin_idx < g_bin_count; bin_idx++) {
+		ret = binary_manager_read_header(bin_idx, &header_data);
+		if (ret != OK) {
+			bmvdbg("Failed to read %dth binary header, errno %d\n", bin_idx, errno);
+			continue;
+		}
+		/* Set the data in table from header */
+		BIN_SIZE(bin_idx) = header_data.bin_size;
+		BIN_OFFSET(bin_idx) = header_data.header_size;
+		strncpy(BIN_VER(bin_idx), header_data.bin_ver, BIN_VER_MAX);
+		strncpy(BIN_KERNEL_VER(bin_idx), header_data.kernel_ver, KERNEL_VER_MAX);
+		strncpy(BIN_NAME(bin_idx), header_data.bin_name, BIN_NAME_MAX);
+		
+		bmvdbg("BIN TABLE[%d] %d %s %s %s\n", bin_idx, BIN_SIZE(bin_idx), BIN_VER(bin_idx), BIN_KERNEL_VER(bin_idx), BIN_NAME(bin_idx));
+	}
+}
+
+static void binary_manager_load_binary(void)
+{
+	int bin_idx;
+	pid_t pid;
+	char devname[BINMGR_DEVNAME_LEN];
+
+	for (bin_idx = 0; bin_idx < g_bin_count; bin_idx++) {
+		snprintf(devname, BINMGR_DEVNAME_LEN, BINMGR_DEVNAME_FMT, BIN_PARTNUM(bin_idx, BIN_USEIDX(bin_idx)));		
+		bmvdbg("BIN[%d] %s %d %d\n", bin_idx, devname, BIN_SIZE(bin_idx), BIN_OFFSET(bin_idx));
+		pid = load_binary(devname, BIN_SIZE(bin_idx), BIN_OFFSET(bin_idx));
+		if (pid <= 0) {
+			bmdbg("Load '%s' fail, errno %d\n", BIN_NAME(bin_idx), errno);
+		} else {
+			BIN_ID(bin_idx) = pid;
+			bmvdbg("Load '%s' success! pid = %d\n", BIN_NAME(bin_idx), BIN_ID(bin_idx));
+		}
+	}
+}
+
+void binary_manager_board_reset(void)
+{
+#ifdef CONFIG_BOARD_ASSERT_AUTORESET
+	boardctl(BOARDIOC_RESET, EXIT_SUCCESS);
+#else
+	(void)irqsave();
+	for (;;) {
+#ifdef CONFIG_ARCH_LEDS
+		//board_autoled_on(LED_PANIC);
```

Comment:
If not needed, please remove this line.

Reply:
I will update it after updating up_assert for arm-M